module/extension/Director/tests/start_stop.cpp

```diff
@@ -0,0 +1,81 @@
+#include <catch.hpp>
+#include <nuclear>
+
+#include "Director.hpp"
+#include "TestBase.hpp"
+
+// Anonymous namespace to avoid name collisions
+namespace {
+
+    struct SimpleTask {};
+    struct Runner {
+        Runner(const bool& task_) : task(task_) {}
+        bool task;
+    };
+
+    std::vector<std::string> events;
+
+    class TestReactor : public TestBase<TestReactor> {
+    public:
+        explicit TestReactor(std::unique_ptr<NUClear::Environment> environment)
+            : TestBase<TestReactor>(std::move(environment)) {
+
+            on<Start<SimpleTask>>().then([this] { events.push_back("start"); });
+            on<Provide<SimpleTask>>().then([this] { events.push_back("provide"); });
+            on<Stop<SimpleTask>>().then([this] { events.push_back("stop"); });
+
+            on<Provide<Runner>>().then([this](const Runner& runner) {
+                // Emit a task to execute
+                if (runner.task) {
+                    events.push_back("running task");
+                    emit<Task>(std::make_unique<SimpleTask>());
+                }
+                else {
+                    events.push_back("removing task");
```

Comment:
Does this stop because it runs but doesn't emit any tasks?

Reply:
When it emitted the task previously that task becomes this provider's "task pack" or pack of functions it wants to run.
The second time when it emits nothing that means it removes all tasks it didn't reemit.

So you can imagine for a provider, each time it runs it emits the tasks it needs.
If it fails to emit one, then the currently running tasks it didn't emit are ended.